Fare integration

antoine-de opened the pull request:
Integration of the fare module in navitia.

fares are computed for each journey.

antoine-de committed:
Fare: initialization

copy of fare files from fare repository

antoine-de committed:
Fare: split parser and algorithm

split all data load in ed/connector.

antoine-de committed:
Fare: SectionKey constructor change

To minimize impact of code, for the moment we try not the change the
algorithm structures, so we only change the way SectionKeys are created.

unfortunatly to do that some refactory to routing::PathItem had to be
done.

Add boost serialization to

For the moment Unit tests are broken :/ so they are commented

antoine-de committed:
Fare: fix UT

departure and arrival were switched...

antoine-de committed:
Fare: cleanup

Commit:
ed: tables added for "fare".

Commit:
fare : Write price informations in database
tables from csv files.

antoine-de committed:
Kraken: change routing::Pahtitem datetime to boost::posixtime

change all datetime in PathItem to boost::posix_time, to ease
use in fare and in raptor_api

antoine-de committed:
Fare: call for fare in raptor_api

fare is now fully integrated in kraken.

When a journey computation is asked, the fare is computed for each
journey and the results are given in the protobuff.

It lacks the links to each section (and all the data import in ED)

antoine-de committed:
Kraken: add links to sections in fare

Fares are now linked to sections in the protobuff response

antoine-de committed:
Kraken: update fare OD parser

refacto of the OD fare parser to have only one method for generic OD
fare and stif OD fare.

*WARNING*: the file format has thus changed.
- 2 column have been added to specify if the O/D applies to a stop area,
a mode or a zone (no more hack in code for Paris)
- the last field of the stiff file was not read (don't ask me why :) ).
ALL field are now read, but the test file has been modified (last col
removed) to maintain the same behavior.

antoine-de committed:
Kraken: Ed fare strucuture change

change of the ed fare structure to be able to do some coherence check
while loading Fare CSV

Now the Ed fare structure is much closer to the Navitia Fare structure
(is almost the same without the graph loaded)

antoine-de committed:
Kraken: fare import in ED

writing connector to ED (reading still to be done)

UT needs to be updated since coherence check on ticket existance break
them :D

antoine-de committed:
Kraken: Fare reader

load fare data from ED + serialization

antoine-de committed:
Kraken: several fare improvement

several fare improvement:
- change global condition to enum
- std::cout hunt
- improve exception log
- the main change is the fix of the serialization :)
the boost::gregorian::period is not default constructible so it was
not possible to serialisation the date ticket (and of course the compil
error was not at all about that :) )
change the date ticket tuple to an object to handle that

antoine-de committed:
Fare: UT are now working with coherence check on tickets!

antoine-de committed:
Jormungandr: integration of the fare results

add fare results to jormungandr response

l-vincent-l commented on source/ed/connectors/fare_parser.cpp:
```diff
@@ -0,0 +1,142 @@
+void fare_parser::load_transitions() {
+    CsvReader reader(state_transition_filename);
+    std::vector<std::string> row;
+    reader.next(); //en-tête
```
En anglais ça serait pas mal

l-vincent-l commented on source/ed/connectors/fare_parser.cpp:
```diff
@@ -0,0 +1,142 @@
+        transition.start_conditions = parse_conditions(row.at(2));
+        transition.end_conditions = parse_conditions(row.at(3));
+        std::vector<std::string> global_conditions;
+        std::string str_condition = boost::algorithm::trim_copy(row.at(4));
```
Okaye, du coup le header ne sert à rien, on estime que les colonnes sont toujours au même endroit ?
ça serait pas mal de mettre au moins un petit todo

l-vincent-l commented on source/ed/connectors/fare_parser.cpp:
```diff
@@ -0,0 +1,142 @@
+}
+
+void fare_parser::load_prices() {
+    CsvReader reader(prices_filename);
```
Y a pas de header sur les prix ?

l-vincent-l commented on source/ed/connectors/fare_parser.cpp:
```diff
@@ -0,0 +1,142 @@
+        }
+        std::string start_saec = boost::algorithm::trim_copy(row[0]);
+        std::string dest_saec = boost::algorithm::trim_copy(row[3]);
+        //col 1 and 3 are the human readable name of the start/end, and are not used
```
0 et 3 non ?

l-vincent-l commented on source/ed/connectors/fare_parser.cpp:
```diff
@@ -0,0 +1,142 @@
+    int count = 0;
+    for (row=reader.next(); !reader.eof(); row = reader.next()) {
+
+        if (row.size() < 7) {
```
les lignes peuvent ne pas toujours avoir la même taille ?

l-vincent-l commented on source/ed/connectors/fare_parser.h:
```diff
@@ -0,0 +1,36 @@
+
+    void load();
+private:
+    /// Charge les deux fichiers obligatoires
```
en anglais plizzzzee :8ball:

l-vincent-l commented on source/ed/ed_persistor.cpp:
```diff
@@ -545,4 +556,135 @@ void EdPersistor::insert_synonyms(const std::map<std::string, std::string>& syno
+        for (const auto& c: transition.end_conditions) {
+            end_cond += c.to_string() + sep;
+            sep = "&";
+        }
```
Pourquoi ne pas avoir utilisé boost join ?

antoine-de replied:
parceque j'ai pas réussi comme un gros low (à faire un truc simple), et que j'avais pas trop de temps, mais si tu me files la ligne je veux bien changer.

l-vincent-l replied:
std::string end_cond = boost::algorithm::join(transition.end_conditions, "&");
if(!end_cond.empty()) { 
    end_cond += "&";
}

Du coup c'est pas forcément mega efficace si tu as besoin d'un & final.

l-vincent-l commented on source/ed/ed_reader.h:
```diff
@@ -57,6 +57,12 @@ struct EdReader{
+    //vectors qui servent pour charger les tarif
+//    std::vector<types::Transition*> transitions;
+//    std::vector<types::Price*> prices;
+//    std::vector<types::Origin_Destination*> origin_destinations;
```
ça m'a l'air d'être du code mort ça non

antoine-de replied:
effectivement merci

l-vincent-l commented on source/ed/types.h:
```diff
@@ -292,4 +292,36 @@ struct StopTime : public Nameable {
+//    std::string gare_arrival;
+//    std::vector<uint32_t> ticket_idx;
+//    std::string delta_zone;
+//};
```
C'est du code mort ?

l-vincent-l commented on source/fare/fare.cpp:
```diff
@@ -0,0 +1,361 @@
+
+        SectionKey section_key(item, section_idx++);
+
+//        LOG4CPLUS_INFO(logger, "Nouvelle section à étudier : " /*<< item */);
```
Code mort ?

antoine-de committed:
Kraken&Jormun: link to unknown fare

When fare has not been found, link to an unknown ticket.

+ several corrections after review
